General performance plan one pager (#11048)
dotnet/msbuild: documentation/specs/proposed/General_perf_onepager.md

# General performance improvements
MSBuild is the main tool used for building various types of projects. It is used by Visual Studio, the .NET CLI, and other build tools. The performance of MSBuild is critical to the productivity of developers. This document outlines our goals to improve overall performance of MSBuild.

## Goals and Motivation

We are aiming for:
 - Searching for opportunities to improve evaluation and build times. We can utilize the data collected by ETW traces, the VS perf lab, and previously identified performance issues.
 - Reducing MSBuild overhead in the Visual Studio IDE.
 - Collecting data to understand the current performance bottlenecks.

This document does not cover specific performance improvements, but rather outlines the general goals and motivation for the performance improvements.

## Impact
    
 - Opening project/solution, branch switching and other operations using MSBuild code in VS should be less impacted by the MSBuild.
 - Overall build times should be reduced. 
 - Even a small improvement can save a lot of time and computing resources across all builds done daily worldwide.

> ## Stakeholders

Reviewer: The formatting in these sections ended up very weird. I think its too many spaces before the list start.

>     - Chet Hust (PM) - as a customer advocate

Reviewer 2: ```suggestion
    - Chet Husk (PM) - as a customer advocate
```

Author: Correcting your own name in a teammate's PR... Sorry, it looks like I need a vacation :)

# General performance improvements

> MSBuild is the main tool used for building .NET projects. It is used by Visual Studio, the .NET CLI, and other build tools. The performance of MSBuild is critical to the productivity of developers. This document outlines our goals to improve overall performance of MSBuild.

Reviewer 3: I wouldn't exclude "C++" nor "SDK projects" in this where that overlaps and falls into the MSBuild "core bucket".

Thanks for this, one more thing I would add would be to ensure that we have:

1) The right telemetry in place to avoid regressions in the wild. We already have this for Visual Studio rebuild, however, we want to fix other scenarios, we should have telemetry in place so that a) we can track that we're making a difference and b) don't regress it.

2) Tests. There's no point fixing performance if someone can make a change very next build and regress it. We have tests in the VS lab, but if we're interested in very specific scenarios, we must have tests running somewhere (our lab can be that if you like) to avoid regressions.

This is how VS does it:

1) Identity scenarios we care about
2) Right tests in lab to track that scenario
3) Have the _right_ telemetry and put on "ShipReady" to monitor and automatically file bugs when it regresses.

Now you folks will be new to a lot of this, and my team can help bring you up to speed how this works and help where needed.

Author: I didn't want to exclude any project type :) The .NET was just on top of my mind. I will write it more generic.

I like your enthusiasm :) We will have a working lab tailored for MSBuild to detect regressions. With the VS lab data, we can cover more scenarios, which is great. Looking forward to join forces and speed up the VS.

## Goals and Motivation

We are aiming for:

>  - Searching opportunities to improve the build times.

Reviewer 3: I would pick build **and** evaluation time. Both are heavily important for Visual Studio.

Author: I considered evaluation as part of the build internally, but I will express it explicitly.

Reviewer 3: We do in-procs evaluations without builds.

>  - Reducing MSBuild overhead in the Visual Studio IDE.

Reviewer 3: I would pick a strawman, such as "Reduce Visual Studio rebuild time at @ 75th by 20%.". We already have telemetry and tracking of this on our performance dashboards on ShipReady. Unlike almost every single one of performance metrics, release over release, rebuild has gotten worse.

Author: It's difficult right now to set a target number when we don't know exactly what is possible to achieve within this scope. I would keep it as a best-effort task, and we can sync on selecting particular improvements based on priorities, such as UI delay, rebuild time, etc.

 - Collecting data to understand the current performance bottlenecks.

This document does not cover specific performance improvements, but rather outlines the general goals and motivation for the performance improvements.

## Impact
    

>     - Opening project/solution, branch switching and other operations using MSBuild code in VS should be less impacted by the MSBuild.
>     - Overall build times should be reduced. 
>     - Even a small improvement can save a lot of time and computing resources across all builds done daily worldwide.

Reviewer 4: Nit: indentation makes the markup interpret this as a code
```suggestion
 - Opening project/solution, branch switching and other operations using MSBuild code in VS should be less impacted by the MSBuild.
 - Overall build times should be reduced. 
 - Even a small improvement can save a lot of time and computing resources across all builds done daily worldwide.
```

Reviewer 4: Same with all the sections below

## Stakeholders

    - Chet Hust (PM) - as a customer advocate
    - David Kean - as a VS performance expert

## Risks

    - Performance improvements might not be as significant as expected.
    - We can break existing functionality while making changes.

>     - We can thow away some ideas due to technical limitations or complexity/improvements ratio.

Reviewer 4: ```suggestion
    - Some ideas and performance measurement findings might need to be skipped due to technical limitations or complexity/improvements ratio.
```

Reviewer 3: One way to limit this is timebox; start with the real easy wins, if you are spending too much time on one win, just move onto the next one until you run out. I'm super confident based on working in this code base, looking at traces and perf in general, we will find more easy wins than we'll be able to fix.

## Cost

Performance improvements are a long-term investment. We need to balance the cost of the improvements with the expected benefits.
We will need to invest time in collecting data, analyzing the data, and implementing improvements.

> Our goal in this scope is to find small and medium size opprotunities. Bigger functionalities such as evaluation caching and RAR caching are described in separate documents.

Reviewer 4: ```suggestion
Our goal in this scope is to find small and medium size opprotunities (Achievable within a single sprint with 1-2 dev investment). Bigger functionalities such as evaluation caching and RAR caching are described in separate documents.
```

## Plan
    

>     - Collect data on the current performance bottlenecks.

Reviewer 4: We might want to name some of the strategies to be used:
 * Top hits in PRISM
 * ETW measurements on batch build / incrmental build and VS build / incremental build on small and mid solutions and detecting opportunities to improve
 * Reviewing some of the earlier identified opportunities: https://github.com/dotnet/msbuild/issues?q=is%3Aissue%20state%3Aopen%20label%3Aperformance

Reviewer 3: Visual Studio can 100% help with this, we have:

1) Performance lab traces with representative samples across almost all project types, including C++, .NET Framework and .NET Core.

2) Real world telemetry of our builds in the wild. I would focus on rebuild where we actually call MSBuildas we mostly try and avoid skipping msbuild when we can due to performance reasons.

3) We are already collecting, or can collect real world traces from machines in the wild for:
   - High CPU usage
   - High GC-time
   - Thread pool starvation
   - Solution load (for evaluation time)
   - Builds (just turning this on, this week)
 
These are analyzed, blamed and then aggregated through our analyzers. These produce bug-level failures that that are ranked by the top issues in the wild. From here we can find very easy low hanging fruit across the entire product. It can, or will identify, big ticket items as well, such as RAR, JITing overhead, lack of Server GC (so GC can't compete), CPU starvation + cost of context switches (due to MSBuild firing up more thread pools than CPUs across all nodes). I suggest we start with the easy, and low hanging fruit and then build the case for big ticket items. I can help with this.

Once we have data, I will come and present and teach you folks how to read this data, download traces, understand the data, etc.
